addons/point_of_sale/models/pos_order.py

```diff
@@ -117,42 +138,41 @@ def _match_payment_to_invoice(self, order):
         order = self.create(self._order_fields(pos_order))
         prec_acc = order.pricelist_id.currency_id.decimal_places
-        journal_ids = set()
+        PosPayment = self.env['pos.payment']
         for payments in pos_order['statement_ids']:
             if not float_is_zero(payments[2]['amount'], precision_digits=prec_acc):
-                order.add_payment(self._payment_fields(payments[2]))
-            journal_ids.add(payments[2]['journal_id'])
+                payment_dict = payments[2]
+                PosPayment.create(dict(pos_order_id=order.id,
+                                        currency_id=order.pricelist_id.currency_id.id,
+                                        **payment_dict))
 
         if pos_session.sequence_number <= pos_order['sequence_number']:
             pos_session.write({'sequence_number': pos_order['sequence_number'] + 1})
             pos_session.refresh()
 
         if not float_is_zero(pos_order['amount_return'], prec_acc):
-            cash_journal_id = pos_session.cash_journal_id.id
-            if not cash_journal_id:
-                # Select for change one of the cash journals used in this
-                # payment
-                cash_journal = self.env['account.journal'].search([
-                    ('type', '=', 'cash'),
-                    ('id', 'in', list(journal_ids)),
-                ], limit=1)
-                if not cash_journal:
-                    # If none, select for change one of the cash journals of the POS
-                    # This is used for example when a customer pays by credit card
-                    # an amount higher than total amount of the order and gets cash back
-                    cash_journal = [statement.journal_id for statement in pos_session.statement_ids if statement.journal_id.type == 'cash']
-                    if not cash_journal:
-                        raise UserError(_("No cash statement found for this session. Unable to record returned cash."))
-                cash_journal_id = cash_journal[0].id
-            order.add_payment({
+            cash_payment_methods = pos_session\
+                                    .payment_method_ids\
+                                    .filtered(lambda pm: pm.is_cash_count)\
+                                    .sorted(key=lambda pm: pm.id)
+            payment_method = None
+            if cash_payment_methods:
+                payment_method = cash_payment_methods[0]
+            else:
+                raise UserError(_("No cash statement found for this session. Unable to record returned cash."))
+            return_payment_dict = {
+                'name': _('return'),
+                'pos_order_id': order.id,
                 'amount': -pos_order['amount_return'],
                 'payment_date': fields.Date.context_today(self),
-                'payment_name': _('return'),
-                'journal': cash_journal_id,
-            })
+                'payment_method_id': payment_method.id,
```

Comment:
cash_payment_methods[0].id

Reply:
I see I have to combine this with the above comment.